Add basic WebGL prototype, supporting point circle series only for now.

tommclaughlan opened the pull request:
The intent here is to put the current state of the WebGL work in the 'webgl' dev branch.

tommclaughlan committed:
Add basic WebGL prototype, supporting point circle series only for now.

tommclaughlan committed:
Add circle point shader utilities for setting common parameters such as fill and
stroke.

tommclaughlan committed:
fix: missing semi-colon.

hanna-greaves commented on packages/d3fc-series/src/webgl/point.js:
```diff
@@ -0,0 +1,83 @@
+
+        const accessor = getAccessors();
+
+        const points = new Float32Array(filteredData.length * 3);
```
nit - please extract this magic number to a `const `with a relevant name

hanna-greaves commented on packages/d3fc-webgl/LICENSE:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2015-2016 Scott Logic Ltd.
```
we should update the copyright to the current year : )

tommclaughlan replied:
This might be better off done separately on all packages to avoid having different licenses scattered throughout the repo.

joelwilliamsmith commented on packages/d3fc-series/src/webgl/point.js:
```diff
@@ -0,0 +1,83 @@
+        const accessor = getAccessors();
+
+        const points = new Float32Array(filteredData.length * 3);
+        let pi = 0;
```
`pi` is a slightly strange name for this variable! I was expecting `let pi = 3.141...`

hanna-greaves commented on packages/d3fc-webgl/src/buffers/uniformBuilder.js:
```diff
@@ -0,0 +1,56 @@
+            case 4:
+                gl.uniform4fv(loc, data);
+                break;
+            }
```
Default warning/error case would be nice

tommclaughlan committed:
fix: missing new line

hanna-greaves commented on packages/d3fc-webgl/src/glDraw.js:
```diff
@@ -0,0 +1,91 @@
+    let decorate = () => {};
+
+    const firstDraw = () => {
+        program.buffers().attribute('aVertexPosition', attributeBuilder(data));
```
`aVertexPosition` is used twice, we should put it into a const, perhaps in a different file so that we could add any other shader name definitions in the same place

tommclaughlan replied:
I'd been toying with the idea of having some kind of metadata for shader snippets - essentially at the moment to use a shader snippet you need to know exactly which attributes/uniform names are required.
This may be a good solution for that.

tommclaughlan replied:
In the meantime I'll just pull the name to a const though.

hanna-greaves commented on packages/d3fc-webgl/src/scale/scaleMapper.js:
```diff
@@ -0,0 +1,32 @@
+    let glScale = null;
+    const scaleCopyString = scale.copy.toString();
+
+    if (scaleCopyString === d3Scale.scaleLinear().copy.toString()) {
```
Could you add a comment here when you get chance, I'm not sure I follow the logic of making all these string copies

tommclaughlan replied:
Sure, effectively it's a bit of a hacky way of telling which type of d3 scale is passed in, since there's no `scale.name` / `scale.type` to compare against. `scale.copy` gives a function which is specific to the scale type - it's horrible but it works for now.

I'm very much eager to replace this with something nicer.

hanna-greaves reviewed (approved):
Very cool, left a few minor comments. Would also recommend lowering the version number to <= 1.0.0 before we decide this is ready

tommclaughlan committed:
Address comments from pr #1294.

tommclaughlan committed:
Bring the defaults closer to SVG/canvas

joelwilliamsmith commented on packages/d3fc-webgl/src/program/programBuilder.js:
```diff
@@ -0,0 +1,129 @@
+        gl.linkProgram(program);
+
+        if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
+            console.log(gl.getProgramInfoLog(program));
```
This will probably need to be removed at some point - is it useful to have it in now? If so, I think we should put a `TODO` next to it to flag for merging into master

tommclaughlan committed:
Add default shader utility functions for the circle series.

joelwilliamsmith commented on packages/d3fc-webgl/src/program/programBuilder.js:
```diff
@@ -0,0 +1,129 @@
+        gl.compileShader(shader);
+
+        if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
+            console.log(gl.getShaderInfoLog(shader));
```
Same as above comment

tommclaughlan committed:
Add webgl to the point series example and bring the shaders close to
parity with SVG/Canvas

tommclaughlan committed:
Throw errors when program/shader compilation fails rather than quietly
logging.

chrisprice commented on packages/d3fc-webgl/src/glDraw.js:
```diff
@@ -0,0 +1,93 @@
+            return xScale;
+        }
+        xScale = args[0];
+        return xScale;
```
`xScale` -> `draw`

chrisprice commented on packages/d3fc-webgl/src/glDraw.js:
```diff
@@ -0,0 +1,93 @@
+            return yScale;
+        }
+        yScale = args[0];
+        return yScale;
```
`yScale` -> `draw`

chrisprice commented on packages/d3fc-webgl/src/program/programBuilder.js:
```diff
@@ -0,0 +1,127 @@
+    let buffers = bufferBuilder();
+
+    const build = (gl) => {
+        if (newProgram(gl, program, vertexShader(), fragmentShader())) {
```
double evaluation of shaders

chrisprice commented on packages/d3fc-webgl/src/program/programBuilder.js:
```diff
@@ -0,0 +1,127 @@
+    const build = (gl) => {
+        if (newProgram(gl, program, vertexShader(), fragmentShader())) {
+            gl.deleteProgram(program);
+            program = createProgram(gl, vertexShader(), fragmentShader());
```
see above

chrisprice commented on packages/d3fc-webgl/src/program/programBuilder.js:
```diff
@@ -0,0 +1,127 @@
+
+        if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
+            gl.deleteProgram(program);
+            throw new Error(`Failed to link program : ${gl.getProgramInfoLog(program)}`);
```
You can't call `getProgramInfoLog` on a deleted program. Also can we include the source in the thrown error

chrisprice commented on packages/d3fc-webgl/src/program/programBuilder.js:
```diff
@@ -0,0 +1,127 @@
+        gl.compileShader(shader);
+
+        if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
+            gl.deleteShader(shader);
```
as above

chrisprice commented on packages/d3fc-webgl/src/glDraw.js:
```diff
@@ -0,0 +1,93 @@
+    let drawFn = firstDraw;
+    const draw = () => {
+        context.viewport(0, 0, context.canvas.width, context.canvas.height);
+        program.apply(xScale);
```
Could we apply the co-ordinate to the scale at this point?

chrisprice commented on packages/d3fc-webgl/src/buffers/attributeBuilder.js:
```diff
@@ -0,0 +1,86 @@
+
+        gl.vertexAttribPointer(
+            location,
+            numComponents || data.length / numElements,
```
If we remove the concept of `numElements` here, would that mean we could get rid of `numElements` globally?

I've just been caught out by the above calculation. Specifically I was creating a buffer like this -

```
var buffer = new Float32Array(data.length);
```

Rather than -

```
var buffer = new Float32Array(data.length * 4);
```

Combined with -

```
draw.program().numElements(data.length);
```

chrisprice commented on packages/d3fc-webgl/src/scale/scaleMapper.js:
```diff
@@ -0,0 +1,34 @@
+import glScaleLog from './glScaleLog';
+import glScalePow from './glScalePow';
+
+export default (scale) => {
```
These will always be full range. Not sure what the alternative is but it feels like we need to record that somewhere

chrisprice commented on packages/d3fc-webgl/src/shaders/vertexShaderSnippets.js:
```diff
@@ -0,0 +1,22 @@
+
+export const point = {
+    header: `varying float vSize;`,
+    body: `vSize = aVertexPosition.z;
```
Is there an advantage to using a 4-component vector to store x, y & size?

tommclaughlan committed:
Fix a couple of issues raised in #1294

tommclaughlan committed:
Introduce 'glPoint', a more focused version of the 'glDraw' component
which specifically targets point series.

tommclaughlan committed:
Remove glDraw - this is superceded by the series specific draw
componenents (e.g. glPoint)

tommclaughlan committed:
We need to start with a fresh program each time point is drawn to avoid
duplicate shader code from the decorate function

tommclaughlan committed:
Merge changes to bring sensible defaults for the point series

tommclaughlan committed:
Remove glDraw

tommclaughlan committed:
Fold the 'numElements' call into the draw call

tommclaughlan committed:
Clean up the relationship between point, glPoint and program. Context is
rebound upwards from program.

We are also assuming from here that we are rebuilding the shaders from
scratch during each decorate - this can be optimised later.

tommclaughlan committed:
Fix an import in point

tommclaughlan committed:
Clean up and bring some more consistency to naming throughout

tommclaughlan committed:
Rename antialias.js to antiAlias.js

tommclaughlan committed:
Further naming consistency

joelwilliamsmith commented on packages/d3fc-webgl/package.json:
```diff
@@ -0,0 +1,26 @@
+  "main": "build/d3fc-webgl.js",
+  "module": "index",
+  "scripts": {
+    "bundle": "d3fc-scripts bundle",
```
Flagging this now - will need to change once master is webgl branch is rebased off master
